sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala

```diff
@@ -128,6 +128,19 @@ case class UnresolvedInlineTable(
   lazy val expressionsResolved: Boolean = rows.forall(_.forall(_.resolved))
 }
 
+/**
+ * An temp resolved inline table that holds all the expressions that were checked for
+ * the right shape and common data types.
+ * This is a preparation step for [[org.apache.spark.sql.catalyst.optimizer.EvalInlineTables]] which
+ * will produce a [[org.apache.spark.sql.catalyst.plans.logical.LocalRelation]]
+ * for this inline table.
+ *
+ * @param output list of column attributes
+ * @param rows expressions for the data rows
+ */
+case class TempResolvedInlineTable(rows: Seq[Seq[Expression]], output: Seq[Attribute])
```

Comment:
it's fully resolved, just not evaluated. Maybe remove tmp from the name?

Reply:
To avoid any perf regression, we can put `rows` in a wrapper class so that catalyst rules won't see them by default. These expressions do not need any more extra resolution.

Reply:
maybe this is over-conservative. Many rules will be skipped if the expression is already resolved. We can add the wrapper class when needed in the future.

Reply:
Yep, I also don't think it is a good idea to be that conservative here. Especially if we want to move in direction of making inline tables more flexible, as Serge pointed out.

Please take a look at the latest iteration.